web-api/src/rank_merge.rs

```diff
@@ -0,0 +1,170 @@
+// Copyright 2023 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::{hash::Hash, ops::AddAssign};
+
+use itertools::Itertools;
+use xayn_web_api_shared::elastic::ScoreMap;
+
+pub(crate) fn normalize_scores<K>(mut scores: ScoreMap<K>) -> ScoreMap<K>
+where
+    K: Eq + Hash,
+{
+    let max_score = scores
+        .values()
+        .max_by(|l, r| l.total_cmp(r))
+        .copied()
+        .unwrap_or_default();
+
+    if max_score != 0. {
+        for score in scores.values_mut() {
+            *score /= max_score;
+        }
+    }
+
+    scores
+}
+
+pub(crate) fn normalize_scores_if_max_gt_1<K>(mut scores: ScoreMap<K>) -> ScoreMap<K>
+where
+    K: Eq + Hash,
+{
+    let max_score = scores
+        .values()
+        .max_by(|l, r| l.total_cmp(r))
+        .copied()
+        .unwrap_or_default()
+        .max(1.0);
+
+    for score in scores.values_mut() {
+        *score /= max_score;
+    }
+
+    scores
+}
+
+pub(crate) fn merge_scores_average_duplicates_only<K>(
+    mut scores_1: ScoreMap<K>,
+    scores_2: ScoreMap<K>,
+) -> ScoreMap<K>
+where
+    K: Eq + Hash,
+{
+    for (key, value) in scores_2 {
+        scores_1
+            .entry(key)
+            .and_modify(|score| *score = (*score + value) / 2.)
+            .or_insert(value);
+    }
+    scores_1
+}
+
+pub(crate) fn merge_scores_weighted<K>(
+    scores: impl IntoIterator<Item = (f32, ScoreMap<K>)>,
+) -> ScoreMap<K>
+where
+    K: Eq + Hash,
+{
+    let weighted = scores.into_iter().flat_map(|(weight, mut scores)| {
+        for score in scores.values_mut() {
+            *score *= weight;
+        }
+        scores
+    });
+    collect_summing_repeated(weighted)
+}
+
+pub const DEFAULT_RRF_K: f32 = 60.;
+
+/// Reciprocal Rank Fusion
+pub fn rrf<K>(k: f32, scores: impl IntoIterator<Item = (f32, ScoreMap<K>)>) -> ScoreMap<K>
+where
+    K: Eq + Hash + Ord,
+{
+    let rrf_scores = scores.into_iter().flat_map(|(weight, scores)| {
+        scores
+            .into_iter()
+            // For testing we want to make sure that in case of s1 == s2 we still get a
+            // deterministic result, for this we use the key ordering for equal scores
+            .sorted_by(|(k1, s1), (k2, s2)| s1.total_cmp(s2).then_with(|| k1.cmp(k2)).reverse())
```

Comment:
To get more deterministic results in tests I made sorting a bit more stable where appropriate.